icu4c/source/i18n/messageformat2_formatter.cpp

```diff
@@ -27,12 +27,30 @@ namespace message2 {
     // -------------------------------------
     // Creates a MessageFormat instance based on the pattern.
 
-    MessageFormatter::Builder& MessageFormatter::Builder::setPattern(const UnicodeString& pat, UParseError& parseError, UErrorCode& errorCode) {
+    void MessageFormatter::Builder::clearState() {
         normalizedInput.remove();
+        delete errors;
+        errors = nullptr;
+    }
+
+    MessageFormatter::Builder& MessageFormatter::Builder::setPattern(const UnicodeString& pat,
+                                                                     UParseError& parseError,
+                                                                     UErrorCode& errorCode) {
+        clearState();
+        // Create errors
+        errors = create<StaticErrors>(StaticErrors(errorCode), errorCode);
+        THIS_ON_ERROR(errorCode);
+
         // Parse the pattern
         MFDataModel::Builder tree(errorCode);
         Parser(pat, tree, *errors, normalizedInput).parse(parseError, errorCode);
 
+        // Fail on syntax errors
+        if (errors->hasSyntaxError()) {
+            errors->checkErrors(errorCode);
+            U_ASSERT(U_FAILURE(errorCode));
```

Comment:
The assertion is checking that the error code was set (by `checkErrors()`); it would fail if there's _no_ syntax error despite `hasSyntaxError()` being true. I'll add a comment explaining that.